consensus/src/lib.rs

```diff
@@ -1,27 +1,1101 @@
+        while !self.is_block_in_main_chain(&block_index) {
+            result.insert(0, block_index.clone());
+            block_index = self.get_previous_block_index(block_index.get_block_id())?;
+        }
+        Ok(result)
+    }
+
+    fn reorganize(
+        &mut self,
+        best_block_id: Option<Id<Block>>,
+        new_block_index: &BlockIndex,
+    ) -> Result<(), BlockError> {
+        let new_chain = self.get_new_chain(new_block_index)?;
+
+        // Disconnect the current chain if it is not a genesis
+        if let Some(ref best_block_id) = best_block_id {
+            let common_ancestor = self.get_previous_block_index(new_chain[0].get_block_id())?;
+            let mut current_ancestor = self
+                .db_tx
+                .get_block_index(best_block_id)?
+                .expect("Can't get block index. Inconsistent DB");
+
+            // Disconnect blocks
+            while self.is_block_in_main_chain(&current_ancestor)
+                && current_ancestor.get_block_id() != common_ancestor.get_block_id()
+            {
+                self.disconnect_tip(&mut current_ancestor)?;
+                current_ancestor =
+                    self.get_previous_block_index(current_ancestor.get_block_id())?;
+            }
+        }
+
+        // Connect the new chain
+        for block_index in new_chain {
+            self.connect_tip(&block_index, &best_block_id)?;
+        }
+        Ok(())
+    }
+
+    fn store_cached_inputs(&mut self, cached_inputs: &CachedInputs) -> Result<(), BlockError> {
+        for (tx_id, tx_index) in cached_inputs.iter() {
+            self.db_tx.set_mainchain_tx_index(tx_id, tx_index)?;
+        }
+        Ok(())
+    }
+
+    fn calculate_indices(
+        &self,
+        block: &Block,
+        tx: &Transaction,
+    ) -> Result<TxMainChainIndex, BlockError> {
+        let enc_block = block.encode();
+        let enc_tx = tx.encode();
+        let offset_tx = enc_block
+            .windows(enc_tx.len())
+            .enumerate()
+            .find_map(|(i, d)| (d == enc_tx).then(|| i))
+            .ok_or(BlockError::Unknown)? as u32;
+
+        let tx_position =
+            TxMainChainPosition::new(&block.get_id().get(), offset_tx as u32, enc_tx.len() as u32);
+
+        assert_eq!(
+            &self.db_tx.get_mainchain_tx_by_position(&tx_position).ok().flatten().expect("Database corrupted! "),
+            tx
+        );
+
+        TxMainChainIndex::new(
+            SpendablePosition::from(tx_position),
+            tx.get_outputs().len() as u32,
+        )
+        .map_err(BlockError::from)
+    }
+
+    fn connect_transactions(&mut self, block: &Block) -> Result<(), BlockError> {
+        let mut cached_inputs = CachedInputs::new();
+        for tx in block.transactions().iter() {
+            for input in tx.get_inputs() {
+                let input_index = input.get_outpoint().get_output_index();
+
+                let mut tx_index = match cached_inputs.get(&tx.get_id()) {
```

Comment:
I wonder if `tx_index` be evaluated outside the input loop? Together with my other comment about the insert call the function would look like this:
```
    fn connect_transactions(&mut self, block: &Block) -> Result<(), BlockError> {
        let mut cached_inputs = CachedInputs::new();
        for tx in block.transactions().iter() {
            let mut tx_index = match cached_inputs.get(&tx.get_id()) {
                Some(tx_index) => tx_index.clone(),
                None => self.calculate_indices(block, tx)?,
            };
            for input in tx.get_inputs() {
                let input_index = input.get_outpoint().get_output_index();

                if input_index >= tx_index.get_output_count() {
                    return Err(BlockError::Unknown);
                }
                // Set each input as spent
                tx_index
                    .spend(input_index, Spender::from(tx.get_id()))
                    .map_err(BlockError::from)?;
            }
            cached_inputs.insert(tx.get_id(), tx_index.clone());
        }
        self.store_cached_inputs(&cached_inputs)?;
        Ok(())
    }
```
Does it make sense or am I missing something crucial?

Reply:
You're right.

Reply:
Fixed